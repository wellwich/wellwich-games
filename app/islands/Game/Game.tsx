import { useEffect, useRef, useState } from 'hono/jsx';

const Game = ({ name }: { name: string }) => {
	const gameArea = useRef<HTMLDivElement>(null);
	const [gameTitle, setGameTitle] = useState('Game');
	let phaserInstance: Phaser.Game | null = null;

	useEffect(() => {
		const gameAreaElement = gameArea.current;
		if (!gameAreaElement) {
			console.error('gameAreaElement is null');
			return;
		}

		const makeGame = async () => {
			try {
				const phaser = await import(`../../games/${name}.ts`);
				phaserInstance = phaser.default(gameAreaElement);
				if (!phaserInstance) {
					console.error('phaserInstance is null');
					return;
				}
				setGameTitle(phaserInstance.config.gameTitle);

				const canvas = gameAreaElement.querySelector('canvas');
				if (canvas) {
					canvas.addEventListener('contextmenu', (e) => e.preventDefault());
				} else {
					console.error('Canvas element not found');
				}
			} catch (error) {
				console.error('Error loading Phaser game:', error);
			}
		};

		makeGame();

		return () => {
			if (phaserInstance) {
				phaserInstance.destroy(true);
				phaserInstance = null;
			}

			const canvas = gameAreaElement.querySelector('canvas');
			if (canvas) {
				canvas.removeEventListener('contextmenu', (e) => e.preventDefault());
			}
		};
	}, [name]);

	return (
		<div>
			<h2 class="text-2xl font-bold">{gameTitle}</h2>
			<div ref={gameArea} />
		</div>
	);
};

export default Game;
