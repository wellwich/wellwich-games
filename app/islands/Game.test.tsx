import { render } from "hono/jsx/dom";
import Game from "./Game";

describe("ゲームコンポーネント", () => {
  it("レンダリングされるべき", async () => {
    const container = document.createElement("div");
    render(<Game name="test-game" />, container);

    // 非同期処理が完了するのを待つ
    await new Promise((resolve) => setTimeout(resolve, 1000));

    expect(container).toMatchSnapshot();
  });

  it("タイトルが正常に表示されるべき", async () => {
    const container = document.createElement("div");
    render(<Game name="test-game" />, container);

    // 非同期処理が完了するのを待つ
    await new Promise((resolve) => setTimeout(resolve, 1000));

    const title = container.querySelector("h2");
    expect(title?.textContent).toBe("テストゲーム");
  });
});
