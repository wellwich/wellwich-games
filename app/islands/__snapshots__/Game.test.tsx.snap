// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ゲームコンポーネント レンダリングされるべき 1`] = `
<div>
  <div>
    <h2
      class="text-2xl font-bold"
    >
      テストゲーム
    </h2>
    <div
      style="width: 100%; height: 100%;"
    >
      <canvas
        height="600"
        style="margin-left: 0px; margin-top: 0px;"
        width="800"
      />
    </div>
  </div>
</div>
`;
